internal/pkg/agent/application/actions/handlers/handler_action_diagnostics.go

```diff
@@ -0,0 +1,204 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package handlers
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"time"
+
+	"github.com/elastic/elastic-agent/internal/pkg/agent/application/coordinator"
+	"github.com/elastic/elastic-agent/internal/pkg/agent/control/v2/client"
+	"github.com/elastic/elastic-agent/internal/pkg/agent/control/v2/cproto"
+	"github.com/elastic/elastic-agent/internal/pkg/core/monitoring/config"
+	"github.com/elastic/elastic-agent/internal/pkg/diagnostics"
+	"github.com/elastic/elastic-agent/internal/pkg/fleetapi"
+	"github.com/elastic/elastic-agent/internal/pkg/fleetapi/acker"
+	"github.com/elastic/elastic-agent/pkg/core/logger"
+
+	"golang.org/x/time/rate"
+)
+
+// ErrRateLimit is the rate limit error that is returned if the handler is ran too often.
+// This may occur if the user sends multiple diagnostics actions to an agent in a short duration
+// or if the agent goes offline and retrieves multiple diagnostics actions.
+// In either case the 1st action will succeed and the others will ack with an the error.
+var ErrRateLimit = fmt.Errorf("rate limit exceeded")
+
+// Uploader is the interface used to upload a diagnostics bundle to fleet-server.
+type Uploader interface {
+	UploadDiagnostics(context.Context, string, string, int64, io.Reader) (string, error)
+}
+
+// Diagnostics is the handler to process Diagnostics actions.
+// When a Diagnostics action is received a full diagnostics bundle is taken and uploaded to fleet-server.
+type Diagnostics struct {
+	log      *logger.Logger
+	coord    *coordinator.Coordinator
+	limiter  *rate.Limiter
+	uploader Uploader
+}
+
+// NewDiagnostics returns a new Diagnostics handler.
+func NewDiagnostics(log *logger.Logger, coord *coordinator.Coordinator, cfg config.Limit, uploader Uploader) *Diagnostics {
+	return &Diagnostics{
+		log:      log,
+		coord:    coord,
+		limiter:  rate.NewLimiter(rate.Every(cfg.Interval), cfg.Burst),
+		uploader: uploader,
+	}
+}
+
+// Handle processes the passed Diagnostics action.
+func (h *Diagnostics) Handle(ctx context.Context, a fleetapi.Action, ack acker.Acker) error {
+	h.log.Debugf("handlerDiagnostics: action '%+v' received", a)
```

Comment:
As you stated in the Upgrade action hook that Lee is working on, you should really make this async. It will block communication and check-in while this is running and its probably going to result in behavior that others are not wanting.